server/teamshandler_test.go

```diff
@@ -0,0 +1,349 @@
+	"github.com/fluidkeys/crypto/openpgp"
+	"github.com/fluidkeys/fluidkeys/assert"
+	"github.com/fluidkeys/fluidkeys/exampledata"
+	"github.com/fluidkeys/fluidkeys/pgpkey"
+	"github.com/gofrs/uuid"
+)
+
+func makeArmoredDetachedSignature(dataToSign []byte, privateKey *pgpkey.PgpKey) (string, error) {
+	outputBuf := bytes.NewBuffer(nil)
+	entity := privateKey.Entity
+
+	err := openpgp.ArmoredDetachSign(outputBuf, &entity, bytes.NewReader(dataToSign), nil)
+	if err != nil {
+		return "", err
+	}
+	return outputBuf.String(), nil
+}
+
+func TestCreateTeamHandler(t *testing.T) {
+
+	unlockedKey, err := pgpkey.LoadFromArmoredEncryptedPrivateKey(
+		exampledata.ExamplePrivateKey4, "test4")
+	signerFingerprint := unlockedKey.Fingerprint()
+
+	assert.ErrorIsNil(t, err)
+
+	duplicateUUID, err := uuid.NewV4() // this is used to test the duplicated case
+	assert.ErrorIsNil(t, err)
+
+	goodUUID, err := uuid.FromString("74bb40b4-3510-11e9-968e-53c38df634be")
+	assert.ErrorIsNil(t, err)
+
+	goodRoster := `
+uuid = "74bb40b4-3510-11e9-968e-53c38df634be"
+
+[[person]]
+email = "test4@example.com"
+fingerprint = "BB3C 44BF 188D 56E6 35F4  A092 F73D 2F05 33D7 F9D6"
+
+[[person]]
+email = "b@example.com"
+fingerprint = "BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB"
+`
+
+	goodSignature, err := makeArmoredDetachedSignature([]byte(goodRoster), unlockedKey)
+	assert.ErrorIsNil(t, err)
+
+	setup := func() {
+		assert.ErrorIsNil(t,
+			datastore.UpsertPublicKey(nil, exampledata.ExamplePublicKey4))
+
+		assert.ErrorIsNil(t,
+			datastore.LinkEmailToFingerprint(
+				nil, "test4@example.com", exampledata.ExampleFingerprint4,
+			),
+		)
+
+		assert.ErrorIsNil(t,
+			datastore.UpsertPublicKey(nil, exampledata.ExamplePublicKey2))
+
+		assert.ErrorIsNil(t,
+			datastore.LinkEmailToFingerprint(
+				nil, "test2@example.com", exampledata.ExampleFingerprint2,
+			),
+		)
+
+	}
+
+	teardown := func() {
+		_, err := datastore.DeletePublicKey(exampledata.ExampleFingerprint4)
+		assert.ErrorIsNil(t, err)
+
+		_, err = datastore.DeleteTeam(nil, goodUUID)
+		assert.ErrorIsNil(t, err)
+
+		_, err = datastore.DeleteTeam(nil, duplicateUUID)
+		assert.ErrorIsNil(t, err)
+	}
+
+	setup()
+	defer teardown()
+
+	t.Run("creates team from valid roster and signature", func(t *testing.T) {
+		requestData := v1structs.UpsertTeamRequest{
+			TeamRoster:               goodRoster,
+			ArmoredDetachedSignature: goodSignature,
+		}
+
+		response := callAPIWithJSON(t, "POST", "/v1/teams", requestData, &signerFingerprint)
+
+		t.Run("returns HTTP 201", func(t *testing.T) {
+			assertStatusCode(t, http.StatusCreated, response.Code)
+		})
+
+		t.Run("adds valid database row", func(t *testing.T) {
+			team, err := datastore.GetTeam(nil, goodUUID)
+			assert.ErrorIsNil(t, err)
+
+			assert.Equal(t, goodUUID, team.UUID)
+			assert.Equal(t, goodRoster, team.Roster)
+			assert.Equal(t, goodSignature, team.RosterSignature)
+
+			if now := time.Now(); team.CreatedAt.Sub(now) > time.Duration(5)*time.Second {
```

Comment:
Hah!

Reply:
:man_facepalming: I couldn't figure out how to mock time *inside* a request handler